Chan.Part.API: channel participation relation and status metrics

wlahti opened the pull request:
#### Type of change

- New feature

#### Description

Add metrics to track the channel participation relation and status per channel for the orderer.


#### Related issues

[FAB-18087](https://jira.hyperledger.org/browse/FAB-18087)

wlahti committed:
Chan.Part.API: channel participation relation/status metrics

FAB-18087

Signed-off-by: Will Lahti <wtlahti@us.ibm.com>

tock-ibm commented:
Generally, I am a bit confused by the design of the metrics here.
 
A channel can have cluster-relation:  `{member, follower, config-tracker, none}`
so why only the boolean `is_member` `is_follower` ?

In addition, the status can be `{onboarding, active, inactive}` (and in the future also `error` or `failed`).

Labeling with `channel` is fine, but labeling with `status` will create |status| x  |clusterRelation| = 12 boolean gauges.
Is this what we want? 

Can't we have a gauge that directly spans the enum values of status? for example:
status.%channel = 0 (=onboarding) / 1 (=active) / 2 (=inactive)  / 3 (=error)

wlahti commented:
Adjusted the metrics and updated the places I was already setting the metrics in this PR. Need to add reporting of the metrics to at least the config-tracker codepath (and consider whether we care to report cluster relation "none" for non-raft orderers).

tock-ibm commented on orderer/common/multichannel/metrics.go:
```diff
@@ -0,0 +1,87 @@
+		r = 3
+	}
+	m.ClusterRelation.With("channel", channel).Set(float64(r))
+}
```
or maybe a more robust approach would be to add to the `types` package a couple of methods that convert the `types.ClusterRelation*` and `type.Status*` into integers?

tock-ibm replied:
How about adding a default here that would catch if we changed/added to the enums?
This way if we change the enum it translates to 0=None... 
say -1? that would get the admins attention.

wlahti replied:
Do you think it's likely we end up with additional cluster relations/statuses? I do think we should have a default, but maybe panic instead stating it's a programming error and a new metric conversion is needed in the switch statement?

I ordered these in ascending order to go along with the highest number being the most active in terms of participating in the cluster/channel, but if we anticipate more, maybe it'd be best to have 0 remain none/inactive, 1 = member/active, and then the rest. That way we keep it stable and there isn't a presumed progression of the values. Thoughts? 

So maybe for cluster relation: 
0 = none
1 = member
2 = follower
3 = config-tracker


And status:
0 = inactive
1 = active
2 = onboarding

tock-ibm replied:
The ordering looks good. I like the fact that <1,1> is the "normal". I agree that panic() is a good solution for default. This way we would hopefully catch inconsistencies in IT.

I think one additional status that is already in the cards is `types.StatusFailure` - see [FAB-18106]( https://jira.hyperledger.org/browse/FAB-18106), and also my comment [here](https://github.com/hyperledger/fabric/pull/2027#discussion_r516476013). 

Maybe adding it (`types.StatusFailure`) in the context of this task is not a bad idea?

tock-ibm commented on orderer/common/multichannel/metrics.go:
```diff
@@ -0,0 +1,87 @@
+		s = 1
+	case types.StatusActive:
+		s = 2
+	}
```
How about adding a default here that would catch if we changed/added to the enums?
This way if we change the enum it translates to 0=inactive... 
say -1? that would get the admins attention.

tock-ibm commented on docs/source/metrics_reference.rst:
```diff
@@ -61,6 +61,12 @@ The following orderer metrics are exported for consumption by Prometheus.
 |                                              |           |                                                            | channel   |                                                                    |
 +----------------------------------------------+-----------+------------------------------------------------------------+-----------+--------------------------------------------------------------------+
+| cluster_relation                             | gauge     | The cluster relation of the node: 0 if none, 1 if member,  | channel   |                                                                    |
+|                                              |           | 2 if follower, 3 if config-tracker.                        |           |                                                                    |
```
"member" -> "consenter"

tock-ibm commented on docs/source/metrics_reference.rst:
```diff
@@ -232,6 +238,12 @@ associated with the metric.
 | cluster.comm.msg_send_time.%{host}.%{channel}                             | histogram | The time it takes to send a message in seconds.            |
 +---------------------------------------------------------------------------+-----------+------------------------------------------------------------+
+| cluster.relation.%{channel}                                               | gauge     | The cluster relation of the node: 0 if none, 1 if member,  |
+|                                                                           |           | 2 if follower, 3 if config-tracker.                        |
```
member -> consenter

tock-ibm commented on orderer/common/multichannel/metrics.go:
```diff
@@ -0,0 +1,92 @@
+		Namespace:    "cluster",
+		Subsystem:    "",
+		Name:         "relation",
+		Help:         "The cluster relation of the node: 0 if none, 1 if member, 2 if follower, 3 if config-tracker.",
```
member -> consenter

wlahti replied:
Done

tock-ibm commented on orderer/common/multichannel/metrics.go:
```diff
@@ -0,0 +1,92 @@
+	switch relation {
+	case types.ClusterRelationNone:
+		r = 0
+	case types.ClusterRelationMember:
```
wasn't this supposed to be `types.ClusterRelationConsenter`? probably needs a rebase as well

wlahti replied:
Done

tock-ibm commented:
Looks good, except that the transition of cluster-relation member->consenter needs to be applied here as well.

wlahti commented:
> Looks good, except that the transition of cluster-relation member->consenter needs to be applied here as well.

Thanks for catching that! Forgot that this PR was also affected by the rename somehow. :)

jyellick commented on orderer/common/follower/follower_chain.go:
```diff
@@ -374,12 +386,15 @@ func (c *Chain) pullUpToJoin() error {
 func (c *Chain) pullAfterJoin() error {
 	c.setStatus(types.StatusActive)
+	clusterRelation, status := c.StatusReport()
+	c.clusterMetricsReporter.ReportClusterRelationAndStatusMetrics(c.ledgerResources.ChannelID(), clusterRelation, status)
```
Why not embed this into `setStatus`?  You need to call it any time the status is updated, and, you won't have to re-acquire the lock that's done in StatusReport().  Also, if someone ever sets status elsewhere, they won't have to remember to update the metric.

wlahti replied:
Good idea. Done (also added to setClusterRelation).

jyellick commented on orderer/common/multichannel/metrics.go:
```diff
@@ -0,0 +1,92 @@
+		Namespace:    "cluster",
+		Subsystem:    "",
+		Name:         "status",
+		Help:         "The status of the cluster node: 0 if inactive, 1 if active, 2 if onboarding.",
```
I'm not totally convinced here, but wouldn't it be more natural to go 0 inactive, 1 onboarding, 2 active? (ie, typically you transition from 0 to 1 to 2?)

wlahti replied:
That's how I originally implemented this but decided against it due to the fact that we may end up adding different options in the future. For the sake of keeping a stable base set of statuses and allowing new ones in the future, I thought it might be best to move away from the concept of a progression in the enum.

jyellick commented on orderer/common/multichannel/metrics.go:
```diff
@@ -0,0 +1,92 @@
+
+var (
+	clusterStatusOpts = metrics.GaugeOpts{
+		Namespace:    "cluster",
```
So, we already have a "cluster" namespace, used by `orderer/common/cluster`, but it does specify a subsystem of `comm`.  Perhaps we should add a subsystem of `participation` here? Or pick a different namespace?

wlahti replied:
I definitely like the idea of adding "participation" as the subsystem. And I wasn't in love with using "cluster" as the namespace. Maybe it should really be "channel"?

wlahti replied:
Went with "cluster" and "participation" for now (and updated the help text) since the channel name is included as a label.

jyellick replied:
The more I think on this, the more convinced I am that 'cluster' isn't right, and I'm a little confused why we call it cluster status at all.  It has nothing to do with the cluster? Being a follower does not make you a member of the cluster... why not a vanilla channel_participation?

wlahti replied:
Are you suggesting we change the clusterRelation in general reported by the channel participation API in addition to the metric? We could rename that to Relation if it's not cluster-related and then the metric could be namepsace "channel" subsystem "participation" and name "relation" for channel_participation_relation.

wlahti replied:
Updated to what I mentioned in my previous comment.

wlahti replied:
Updated to "participation" for the namespace and no subsystem.

jyellick commented on orderer/common/multichannel/metrics.go:
```diff
@@ -0,0 +1,92 @@
+	clusterStatusOpts = metrics.GaugeOpts{
+		Namespace:    "cluster",
+		Subsystem:    "",
+		Name:         "status",
```
The name `status` is pretty hopelessly vague, unless we add more context with a subsystem or different namespace.  A cluster status of 'active' for instance does not imply that the other nodes are reachable, etc.  It is the `participation_status` or similar.

wlahti replied:
Done. It's participation_status now with the namespace change.

jyellick commented on orderer/common/multichannel/metrics.go:
```diff
@@ -0,0 +1,92 @@
+	clusterRelationOpts = metrics.GaugeOpts{
+		Namespace:    "cluster",
+		Subsystem:    "",
+		Name:         "relation",
```
I also wonder if `relation` is the best we can do here?  I understand that's what it's referred to in the code, but it's not exactly obvious to someone looking at a metrics dashboard.

wlahti replied:
Created follow-up task https://jira.hyperledger.org/browse/FAB-18327 for renaming this (and the underlying types.ClusterRelation) to consensus relation.

jyellick commented on orderer/consensus/etcdraft/consenter.go:
```diff
@@ -160,10 +162,12 @@ func (c *Consenter) HandleChain(support consensus.ConsenterSupport, metadata *co
 			c.InactiveChainRegistry.TrackChain(support.ChannelID(), support.Block(0), func() {
 				c.ChainManager.CreateChain(support.ChannelID())
 			})
+			c.ChainManager.ReportClusterRelationAndStatusMetrics(support.ChannelID(), types.ClusterRelationConfigTracker, types.StatusInactive)
```
I _really_ do not like this.  And this is one of the strong reasons why I asked that we please ditch the notion of the raft consenter switching the chain out, and delegate this back up to the multichannel registrar.  Consensus plugins have no business setting cluster relation status.

wlahti replied:
Yes, I did not like doing this but it's the only way I found with the current code flow.

jyellick replied:
Do we have an outstanding JIRA to fix this flow?

wlahti replied:
No, I do not see a JIRA for this. I can try to create one but I'm not sure if I have enough context in this area to get it completely right.

tock-ibm replied:
I believe there is a Jira for it, or a variation on this theme: https://jira.hyperledger.org/browse/FAB-18152

However, the big landmine we have to dismantle is the pattern with the `InactiveChainRegistry` etc., and we decided back then to finish the channel participation functionality first, and only then start refactoring. Moreover, the dependencies around the `InactiveChainRegistry` and `BlockPuller` are pretty complex, so this should be done very carefully... (if at all - "if it ain't broke don't fix it" comes to mind). Hopefully when we eventually deprecate the system channel this task will be a lot easier.
